mb_mgr_code.h

```diff
@@ -1454,49 +1453,55 @@ is_job_invalid(const JOB_AES_HMAC *job)
                 INVALID_PRN("hash_alg:%d\n", job->hash_alg);
                 return 1;
         }
+        return 0;
+}
 
-        switch (job->chain_order) {
-        case CIPHER_HASH:
-                if (job->cipher_direction != ENCRYPT) {
-                        INVALID_PRN("chain_order:%d\n", job->chain_order);
-                        return 1;
-                }
-                break;
-        case HASH_CIPHER:
-                if (job->cipher_mode != NULL_CIPHER) {
-                        if (job->cipher_direction != DECRYPT) {
-                                INVALID_PRN("chain_order:%d\n",
-                                            job->chain_order);
-                                return 1;
-                        }
-                }
-                break;
-        default:
-                INVALID_PRN("chain_order:%d\n", job->chain_order);
-                return 1;
-        }
+__forceinline
+JOB_AES_HMAC *SUBMIT_JOB_AES(MB_MGR *state, JOB_AES_HMAC *job)
+{
+	if (job->cipher_direction == ENCRYPT)
+		job = SUBMIT_JOB_AES_ENC(state, job);
+	else
+		job = SUBMIT_JOB_AES_DEC(state, job);
 
-        return 0;
+	return job;
+}
+
+__forceinline
+JOB_AES_HMAC *FLUSH_JOB_AES(MB_MGR *state, JOB_AES_HMAC *job)
+{
+	if (job->cipher_direction == ENCRYPT)
+		job = FLUSH_JOB_AES_ENC(state, job);
+	else
+		job = FLUSH_JOB_AES_DEC(state, job);
+
+	return job;
+}
+
+/* submit a half-completed job, based on the status */
+__forceinline
+JOB_AES_HMAC *RESUBMIT_JOB(MB_MGR *state, JOB_AES_HMAC *job)
+{
+	while (job != NULL && job->status < STS_COMPLETED) {
```

Comment:
Use of 'while' here will be certainly a problem for the submit_job() path. 'if' may just be fine - any particular reason while loop is used here?

Reply:
actually, with the addition of new operation order this 'while' turns to be absolutely crucial